### PR TITLE
Type holes

### PR DESCRIPTION
**Pull request by akrmn**
# Type holes

This PR adds support for type holes (`?`, `?foo`, `?bar`). These are only allowed in stack assertions. Type holes are traced when unified with another type, which is useful for understanding what types are on the stack at a given point.

In action:
```mirth
def(some-function, a -- b +IO,
  ...
  { ?a ?b }
  ...
)
```
will trace
```
?a ~ Int
?b ~ Bool
```
if the stack has `[Int Bool]` at that point of the program.

# Type Don't Care

It also adds support for a "don't care" type, `_`. This is also only allowed in stack assertions, but it is not traced. This is useful for asserting some types on the stack while ignoring others.

In action:
```mirth
def(some-other-function, a -- b +IO,
  ...
  { Bool _ Int _ }
  ...
)
```
will assert that, at that point of the program, the second element on the stack is an `Int` and the fourth a `Bool`, while asserting nothing about the topmost element or the third element from the top.

### CODE REVIEW

**Comment by akrmn** [comment]
hm maybe the `_(foo)` syntax isn't the best if we want to allow higher-order type holes, i.e. `_(Int)`. perhaps a Haskellish `_foo` would be a better fit.

**Comment by akrmn** [comment]
Hi! I updated the description of the PR. The main change is that type holes are now introduced with a question mark, and no special distinction is made between named type holes (`?foo`, `?bar`) and the anonymous type hole `?`. Additionally, the old anonymous type hole, `_`, is internally renamed to `TYPE_DONT_CARE`, and doesn't take any arguments nor allow any chars after the underscore.

Also, I tweaked the check on where type holes are allowed. Previously, it was an up-front check when elaborating the signature for a word definition. This had the problem that if a new function had to elaborate a signature, the author had to remember to add this check. Now, the check is done when a type hole/type don't care is found, with the help of an argument `HolesAllowed`, which must be passed explicitly, but is for the most part ignored.

**Comment by sofiafaro** [comment]
Fantastic, thank you! I'm going to review it carefully :-)

File: src/mirth/elab.mth [hunk 1]
```diff
@@ -109,7 +113,30 @@ def(elab-type-con!, Ctx Token -- Ctx Type Token,
...
+        token-location "type holes are not allowed here" emit-fatal-error-at!
+    ))
+
+def(elab-type-dont-care!, HolesAllowed Ctx Token -- HolesAllowed Ctx Type Token,
+    dip(dip(dup)) rotl HolesAllowed->Bool if(
+        token-has-args? if(
+            dup "Types with args not yet supported." emit-error! TYPE_ERROR,
+            TYPE_DONT_CARE)
+        swap token-next,
+        token-location "type don't care is not allowed here" emit-fatal-error-at!
```
Comment (sofiafaro) [inline src/mirth/elab.mth:136]:
```suggestion
        "type don't care is not allowed here" emit-fatal-error!
```

File: src/mirth/elab.mth [hunk 2]
```diff
@@ -109,7 +113,30 @@ def(elab-type-con!, Ctx Token -- Ctx Type Token,
...
+def(allow-holes, HolesAllowed, true cast)
+def(forbid-holes, HolesAllowed, true cast)
+
+def(elab-type-hole!, HolesAllowed Ctx Token -- HolesAllowed Ctx Type Token,
+    dip(dip(dup)) rotl HolesAllowed->Bool if(
+        token-has-args? if(
+            dup "Types with args not yet supported." emit-error! TYPE_ERROR,
+            token-name? THole)
+        swap token-next,
+        token-location "type holes are not allowed here" emit-fatal-error-at!
```
Comment (sofiafaro) [inline src/mirth/elab.mth:127]:
```suggestion
        "type holes are not allowed here" emit-fatal-error!
```

File: src/mirth/data/name.mth [hunk 3]
```diff
@@ -159,14 +159,32 @@ def(name-could-be-type, Name -- Bool,
...
+def(name-is-type-dont-care, Name -- Bool,
+    name-quads& dup u8@ is-underscore? nip if(
+        str-tail str-null,
+        drop false
+    ))
+
 def(name-could-be-stack-var, Name -- Bool,
     name-quads& dup u8@ is-asterisk? nip if(
-        str-tail u8@ is-lower? nip,
+        str-tail str-could-be-type-var,
```
[the remark this answers is not in the document]
Comment (akrmn) [reply src/mirth/data/name.mth:187]: thanks!

File: src/mirth/elab.mth (hunk 2 above)
Comment (akrmn) [reply src/mirth/elab.mth:127]: oh! thanks :)

**Comment by akrmn** [comment]
thanks for the suggestions!

**Pull request merged**